ompi/util/timings.h

```diff
@@ -4,123 +4,215 @@
 #include "opal/util/timings.h"
 /* TODO: we need access to MPI_* functions */
 
-#if (0 && OPAL_ENABLE_TIMING)
-
-/* TODO: replace with opal_timing function */
-static inline double OMPI_TIMING_GET_TS(void)
-{
-    struct timespec ts;
-    double ret;
-    clock_gettime(CLOCK_MONOTONIC, &ts);
-    ret = ts.tv_sec + 1E-9 * ts.tv_nsec;
-    return ret;
+#if (OPAL_ENABLE_TIMING)
+
+typedef struct {
+    char **desc;
+    double *in;
+    int use;
+} ompi_timing_item_t;
+
+typedef struct ompi_timing_list_t {
+    ompi_timing_item_t *item;
+    struct ompi_timing_list_t *next;
+} ompi_timing_list_t;
+
+typedef struct {
+    double ts;
+    const char *prefix;
+    int inum;
+    int cnt;
+    int error;
+    int enabled;
+    opal_timing_ts_func_t get_ts;
+    ompi_timing_list_t *timing;
+} ompi_timing_env_t;
+
+#define OMPI_TIMIG_ENV_ITEM_IDX(idx) ({                                        \
+    __label__ exit;                                                            \
```

Comment:
This is GCC extension, it allows to declare local labels. More info: https://gcc.gnu.org/onlinedocs/gcc/Local-Labels.html